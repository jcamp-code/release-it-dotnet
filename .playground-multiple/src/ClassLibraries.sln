
Microsoft Visual Studio Solution File, Format Version 12.00
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ClassLibrary1", "ClassLibrary1\ClassLibrary1.csproj", "{157EC7C0-5249-4663-86F6-BD53CFEADE03}"
EndProject
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ClassLibrary2", "ClassLibrary2\ClassLibrary2.csproj", "{BE11FE1F-AA61-4E7C-805A-B93518C2ABA5}"
EndProject
Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{0213EEBE-4810-46C4-A27E-DE459BAE076A}"
	ProjectSection(SolutionItems) = preProject
		Directory.Build.props = Directory.Build.props
		..\.release-it.cjs = ..\.release-it.cjs
	EndProjectSection
EndProject
Global
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|Any CPU = Debug|Any CPU
		Release|Any CPU = Release|Any CPU
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{157EC7C0-5249-4663-86F6-BD53CFEADE03}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{157EC7C0-5249-4663-86F6-BD53CFEADE03}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{157EC7C0-5249-4663-86F6-BD53CFEADE03}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{157EC7C0-5249-4663-86F6-BD53CFEADE03}.Release|Any CPU.Build.0 = Release|Any CPU
		{BE11FE1F-AA61-4E7C-805A-B93518C2ABA5}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{BE11FE1F-AA61-4E7C-805A-B93518C2ABA5}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{BE11FE1F-AA61-4E7C-805A-B93518C2ABA5}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{BE11FE1F-AA61-4E7C-805A-B93518C2ABA5}.Release|Any CPU.Build.0 = Release|Any CPU
	EndGlobalSection
EndGlobal
