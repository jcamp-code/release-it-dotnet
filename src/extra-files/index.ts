import { existsSync, promises as fsp } from 'node:fs'
import { globby, isDynamicPattern } from 'globby'

import { Generic } from './rp/generic'
import { GenericJson } from './rp/generic-json'
import { GenericXml } from './rp/generic-xml'
import { GenericYaml } from './rp/generic-yaml'
import type { Update, Updater } from './rp/update'
import { Version } from './rp/version'

export * from './knownFileTypes'

export interface ExtraJsonFile {
  type: 'json'
  path: string
  jsonpath: string
}
export interface ExtraYamlFile {
  type: 'yaml'
  path: string
  jsonpath: string
}
export interface ExtraXmlFile {
  type: 'xml'
  path: string
  xpath: string
}
export type ExtraFile = string | ExtraJsonFile | ExtraYamlFile | ExtraXmlFile

async function extraFilePaths(extraFile: ExtraFile): Promise<string[]> {
  const path = typeof extraFile === 'string' ? extraFile : extraFile.path
  if (!isDynamicPattern(path)) return [path]
  return await globby(path)
}

function getUpdate(
  type: string,
  path: string,
  updatePath: string,
  version: Version,
): Update {
  switch (type) {
    case 'json':
      return {
        path,
        createIfMissing: false,
        updater: new GenericJson(updatePath, version),
      }
    case 'yaml':
      return {
        path,
        createIfMissing: false,
        updater: new GenericYaml(updatePath, version),
      }
    case 'xml':
      return {
        path,
        createIfMissing: false,
        updater: new GenericXml(updatePath, version),
      }
    default:
      throw new Error(`unsupported extraFile type: ${type}`)
  }
}

export async function extraFileUpdates(
  extraFiles: ExtraFile[],
  version: Version,
  knownDefaults?: (string) => { type: string; updatePath: string } | undefined,
): Promise<Update[]> {
  const extraFileUpdates: Update[] = []
  for (const extraFile of extraFiles)
    if (typeof extraFile === 'object') {
      const paths = await extraFilePaths(extraFile)
      for (const path of paths)
        switch (extraFile.type) {
          case 'json':
          case 'yaml':
            extraFileUpdates.push(
              getUpdate(extraFile.type, path, extraFile.jsonpath, version),
            )
            break
          case 'xml':
            extraFileUpdates.push(
              getUpdate('xml', path, extraFile.xpath, version),
            )
            break
          default:
            throw new Error(
              `unsupported extraFile type: ${
                (extraFile as { type: string }).type
              }`,
            )
        }
    } else {
      const paths = await extraFilePaths(extraFile)
      for (const path of paths) {
        if (knownDefaults) {
          const { type, updatePath } = knownDefaults(path)
          if (type && updatePath) {
            extraFileUpdates.push(getUpdate(type, path, updatePath, version))
            continue
          }
        }
        extraFileUpdates.push({
          path,
          createIfMissing: false,
          updater: new Generic({ version }),
        })
      }
    }

  return extraFileUpdates
}

export async function processExtraFiles(
  extraFiles: ExtraFile[],
  version: string,
  isDryRun: boolean,
  knownDefaults?: (string) => { type: string; updatePath: string } | undefined,
): Promise<Update[]> {
  const versionObject = Version.parse(version)
  const updates = await extraFileUpdates(
    extraFiles,
    versionObject,
    knownDefaults,
  )
  for (const update of updates)
    if (existsSync(update.path)) {
      const content = await fsp.readFile(update.path, 'utf-8')
      update.newContent = update.updater.updateContent(content)
      if (!isDryRun) await fsp.writeFile(update.path, update.newContent)
    }

  return updates
}
