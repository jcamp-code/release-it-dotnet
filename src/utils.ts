import { existsSync, promises as fsp, readFileSync } from 'node:fs'
import * as dom from '@xmldom/xmldom'
import * as xpath from 'xpath'

export const readJSON = (file) => JSON.parse(readFileSync(file, 'utf8'))

export async function getXmlVersion(path: string) {
  if (path)
    if (existsSync(path)) {
      const content = await fsp.readFile(path, 'utf-8')
      const document = new dom.DOMParser().parseFromString(content)
      const nodes = xpath.select(
        '//Project/PropertyGroup/Version',
        document,
      ) as Node[]
      if (nodes && nodes[0]) return nodes[0].textContent
    }
}
