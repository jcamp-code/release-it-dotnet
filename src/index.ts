import { promises as fsp } from 'node:fs'
import { cwd, env } from 'node:process'
import { defu } from 'defu'
import { globby } from 'globby'
import { basename, resolve } from 'pathe'
import { Plugin } from 'release-it'
import { rimraf } from 'rimraf'
import shelljs from 'shelljs'

import 'dotenv/config'

import {
  knownDotNetFiles,
  processExtraFiles,
  type ExtraFile,
} from './extra-files'
import { getXmlVersion } from './utils'

const DEFAULT_NUGET_FEED_URL = 'https://api.nuget.org/v3/index.json'

export interface DotNetPluginContext extends DotNetPluginOptions {
  version?: string
  nugetSourceName: string
  bumpedVersion: string
  isPublished: boolean
  cwd?: string
}
export interface DotNetPluginOptions {
  nugetFeedUrl?: string
  nugetApiKey?: string
  packageId?: string
  csprojFile?: string
  versionFile?: string | false
  build?: boolean
  pack: boolean
  publish: boolean
  buildConfiguration: string
  keepArtifacts: boolean
  extraFiles?: string | string[]
}

function getFeedName(nugetFeedUrl: string) {
  if (nugetFeedUrl === null || nugetFeedUrl === DEFAULT_NUGET_FEED_URL)
    return 'NuGet'
  return nugetFeedUrl.replace('https://', '').replace('/v3/index.json', '')
}

const prompts = {
  publish: {
    type: 'confirm',
    message: (context) => {
      const feed = getFeedName(context.nugetFeedUrl)
      return `Publish v${context.version} of ${context.packageId} to ${feed}?`
    },
  },
  build: {
    type: 'confirm',
    message: () => {
      return `Build latest version?`
    },
  },
}

const defaults: DotNetPluginContext = {
  nugetApiKey: env.NUGET_TOKEN || env.NUGET_API_KEY,
  nugetFeedUrl: DEFAULT_NUGET_FEED_URL,
  pack: false,
  build: false,
  publish: true,
  buildConfiguration: 'Release',
  nugetSourceName: 'ReleaseItSource',
  bumpedVersion: '',
  keepArtifacts: false,
  isPublished: false,
}

class DotNetPlugin extends Plugin {
  context: DotNetPluginContext = {} as DotNetPluginContext

  constructor(...args) {
    super(...args)
    this.registerPrompts(prompts)
  }

  getInitialOptions(options, namespace) {
    return options[namespace]
  }

  init() {
    this.context = defu(this.options, defaults)
    this.context.cwd = cwd()

    // can pack without publish, but not the other way around
    if (this.context.publish) this.context.pack = true

    // set version file to csprojFile unless version is explicity set to false
    if (
      this.context.versionFile === undefined &&
      this.context.csprojFile.endsWith('.csproj')
    )
      this.context.versionFile = this.context.csprojFile

    // check the config and environment
    if (this.context.pack || this.context.publish) {
      const dotnetExecutable = shelljs.which('dotnet')
      if (!dotnetExecutable) throw new Error('dotnet executable not found')
      if (!this.context.csprojFile && !this.context.versionFile)
        throw new Error('versionFile or csprojFile is required')

      if (
        this.context.csprojFile &&
        !this.context.versionFile &&
        this.context.csprojFile.endsWith('.sln')
      )
        throw new Error('versionFile required if csprojFile is a solution file')

      if (this.context.publish) {
        if (!this.context.nugetApiKey)
          throw new Error('nugetApiKey is required if publishing is enabled')
        if (!this.context.nugetFeedUrl)
          throw new Error('nugetFeedUrl is required if publishing is enabled')
      }
    }
  }

  static disablePlugin(options) {
    return options.versionFile ? 'npm' : null
  }

  async getLatestVersion() {
    if (this.context?.versionFile) {
      const file = resolve(cwd(), this.context?.versionFile)
      const latestVersion = getXmlVersion(file)
      if (latestVersion === undefined)
        throw new Error(
          `DotNetPlugin: Could not find version info in ${this.context.versionFile}`,
        )
      return latestVersion
    }
  }

  async bump(version) {
    this.context.bumpedVersion = version
  }

  async beforeRelease() {
    const updates: ExtraFile[] = []
    if (this.context?.versionFile) updates.push(this.context?.versionFile)
    if (typeof this.context.extraFiles === 'string')
      updates.push(this.context.extraFiles)
    if (Array.isArray(this.context.extraFiles))
      updates.push(...this.context.extraFiles)

    if (updates.length > 0) {
      this.log.log('🔧 Processing Version Updates...')
      this.log.verbose(updates)

      await processExtraFiles(
        updates,
        this.context.bumpedVersion,
        false,
        knownDotNetFiles,
      )
    }
  }

  async release() {
    const { publish, pack, build } = this.context

    if (build)
      await this.step({
        task: () => this.dotnetBuild(),
        label: 'Build with dotnet',
        prompt: 'build',
      })

    if (!pack && !publish) return

    try {
      await this.dotnetAddSource()
      await this.dotnetPack()

      if (publish) {
        await this.getPackageId()

        await this.step({
          task: () => this.dotnetPublish(),
          label: 'Publish with dotnet nuget',
          prompt: 'publish',
        })
      }
    } catch (e) {
      await this.dotnetCleanUp()
      throw e
    }
  }

  async afterRelease() {
    const feed = getFeedName(this.context.nugetFeedUrl)

    if (this.context.isPublished)
      this.log.log(`🚀 Published v${this.context.bumpedVersion} to ${feed}`)
    await this.dotnetCleanUp()
  }

  // Internal DotNet / Nuget Utils

  async getPackageId() {
    let packageId = this.context.packageId

    if (!packageId) {
      const { csprojFile } = this.context
      if (csprojFile.endsWith('.sln')) {
        packageId = basename(csprojFile).replace('.sln', '')
      } else {
        const file = resolve(cwd(), csprojFile)
        const csprojFileContents = await fsp.readFile(file, 'utf8')
        const match = csprojFileContents.match(/<PackageId>(.*)<\/PackageId>/)
        const packageIdExists = match && match.length > 1
        packageId = packageIdExists
          ? match[1]
          : basename(csprojFile).replace('.csproj', '')
      }
      this.context.packageId = packageId
    }

    this.config.setContext({
      packageId,
      nugetFeedUrl: this.context.nugetFeedUrl,
    })
  }

  async dotnetAddSource() {
    const { nugetFeedUrl, nugetApiKey, nugetSourceName } = this.context

    const log = await this.exec(`dotnet nuget list source`)

    if (log?.includes(nugetFeedUrl)) return

    const command = [
      'dotnet',
      'nuget',
      'add',
      'source',
      nugetFeedUrl,
      '-n',
      nugetSourceName,
      '-u',
      'ReleaseItUser',
      '-p',
      nugetApiKey,
      '--store-password-in-clear-text',
    ]

    this.log.verbose(`DotNetPlugin: Add Source Command: ${command}`)

    await this.exec(command)
  }

  async dotnetBuild() {
    const { csprojFile, buildConfiguration } = this.context

    this.log.log(`🚧 Building ${csprojFile || ''}...`)

    const command = [
      'dotnet',
      'build',
      `"${csprojFile}"`,
      '--configuration',
      buildConfiguration,
    ]

    this.log.verbose(`DotNetPlugin: Building Command: ${command}`)

    await this.exec(command)
  }

  async dotnetPack() {
    const { csprojFile, buildConfiguration, packageId } = this.context

    this.log.log(`🚧 Packing ${csprojFile || packageId}...`)

    const command = [
      'dotnet',
      'pack',
      `"${csprojFile}"`,
      '--configuration',
      buildConfiguration,
      '--output',
      './artifacts',
      // this should be set in the version files instead
      // `-p:PackageVersion=${bumpedVersion}`,
    ]

    this.log.verbose(`DotNetPlugin: Packing Command: ${command}`)

    await this.exec(command)
  }

  async dotnetPublish() {
    try {
      const { nugetFeedUrl, nugetApiKey, packageId } = this.context

      this.log.log(`🚧 Publishing ${packageId}...`)

      const command = [
        'dotnet',
        'nuget',
        'push',
        // changed to publish all packages
        `./artifacts/**.nupkg`,
        // `./artifacts/${packageId}.${bumpedVersion}.nupkg`,
      ]

      command.push('-s')
      command.push(nugetFeedUrl)

      command.push('-k')
      command.push(nugetApiKey)

      this.log.verbose(`DotNetPlugin: Publish Command: ${command}`)

      await this.exec(command, { external: true, write: true })
      this.context.isPublished = true
    } catch (e) {
      await this.dotnetCleanUp()
      throw e
    }
  }

  async dotnetRemoveSource() {
    const { nugetSourceName } = this.context

    const log = await this.exec(`dotnet nuget list source`)

    if (log?.includes(nugetSourceName))
      await this.exec(`dotnet nuget remove source ${nugetSourceName}`)
  }

  async dotnetCleanUp() {
    await this.dotnetRemoveSource()
    if (!this.context.keepArtifacts) {
      await rimraf('./artifacts')
      this.log.log('🗑️ Cleaned up')
    } else {
      const artifacts = await globby(resolve(cwd(), './artifacts/*'))
      this.log.log('📦️ Artifacts Generated')
      this.log.verbose(
        artifacts.filter((artifact) =>
          artifact.includes(this.context.bumpedVersion),
        ),
      )
    }
  }
}
export default DotNetPlugin
