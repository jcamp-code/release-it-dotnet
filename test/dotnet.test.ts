import { promises as fsp } from 'node:fs'
import { factory, runTasks } from 'release-it/test/util/index.js'
import shelljs from 'shelljs'
import { describe, expect, it } from 'vitest'

import Plugin from '../src'
import { GenericXml } from '../src/extra-files/rp/generic-xml'
import { Version } from '../src/extra-files/rp/version'
import { getXmlVersion } from '../src/utils'

const namespace = 'dotnet-plugin'
describe('dotnet-plugin', async () => {
  it('should not throw if publish off', async () => {
    const options = { [namespace]: { publish: false } }
    const plugin = factory(Plugin, { namespace, options })
    expect(async () => {
      await runTasks(plugin)
      // console.log(result)
    }).not.toThrowError()
  })
  it.fails('should throw if no csprojfile', async () => {
    const options = { [namespace]: {} }
    const plugin = factory(Plugin, { namespace, options })
    await expect(runTasks(plugin)).rejects.toBe(1)
  })
  it('updates CS Proj file', async () => {
    const document = await fsp.readFile('./test/fixtures/test.csproj', 'utf-8')
    const xmlUpdater = new GenericXml(
      '//project/propertygroup/version',
      new Version(3, 0, 7),
    )
    expect(xmlUpdater.updateContent(document)).toMatchSnapshot()
  })
  it('updates Directory Props', async () => {
    const document = await fsp.readFile(
      './test/fixtures/Directory.Build.props',
      'utf-8',
    )
    const xmlUpdater = new GenericXml(
      '//project/propertygroup/version',
      new Version(3, 0, 7),
    )
    expect(xmlUpdater.updateContent(document)).toMatchInlineSnapshot(`
      "<Project>

        <PropertyGroup>
          <RootNamespace>MongoFramework.AspNetCore.Identity</RootNamespace>
          <PackageId>$(AssemblyName)</PackageId>
          <PackageRequireLicenseAcceptance>false</PackageRequireLicenseAcceptance>
          <PackageLicenseFile>LICENSE</PackageLicenseFile>
          <PackageProjectUrl>https://github.com/JohnCampionJr/MongoFramework.AspNetCore.Identity</PackageProjectUrl>
          <PackageBaseTags>mongodb;mongoframework;mongo;identity</PackageBaseTags>

          <RepositoryUrl>https://github.com/jcamp-code/release-it</RepositoryUrl>
          <RepositoryType>git</RepositoryType>
          <RepositoryBranch>main</RepositoryBranch>

          <!-- SourceLink Support -->
          <PublishRepositoryUrl>true</PublishRepositoryUrl>
          <EmbedUntrackedSources>true</EmbedUntrackedSources>
          <IncludeSymbols>true</IncludeSymbols>
          <SymbolPackageFormat>snupkg</SymbolPackageFormat>
          <LangVersion>Latest</LangVersion>
          <Version>3.0.7</Version>
        </PropertyGroup>
        <ItemGroup>
          <PackageReference Include=\\"Microsoft.SourceLink.GitHub\\" Version=\\"1.1.1\\" PrivateAssets=\\"All\\"/>
        </ItemGroup>

        <ItemGroup>
          <None Include=\\"..\\\\..\\\\LICENSE\\" Pack=\\"true\\" Visible=\\"false\\" PackagePath=\\"\\"/>
          <None Include=\\"..\\\\..\\\\README.md\\" Pack=\\"true\\" Visible=\\"false\\" PackagePath=\\"\\"/>
        </ItemGroup>

      </Project>
      "
    `)
  })
  it('gets latest version from .csproj', async () => {
    expect(await getXmlVersion('./test/fixtures/test.csproj')).toBe('2.9.7')
  })
  it('verifies present of dotnet', async () => {
    expect(await shelljs.which('dotnet')).toBeTruthy()
  })
})
