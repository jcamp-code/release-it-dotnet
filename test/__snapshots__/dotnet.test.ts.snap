// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dotnet-plugin > updates CS Proj file 1`] = `
"<Project Sdk=\\"Microsoft.NET.Sdk\\">

  <PropertyGroup>
    <TargetFrameworks>net6.0;net8.0</TargetFrameworks>
    <PackageTags>$(PackageBaseTags)</PackageTags>
    <AssemblyName>MyAssembly</AssemblyName>
    <Title>Package Title</Title>
    <PackageTags>$(PackageBaseTags)</PackageTags>
    <Authors>John Campion Jr</Authors>
    <Version>3.0.7</Version>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include=\\"MongoFramework\\" Version=\\"0.29.0\\"/>
    <PackageReference Include=\\"Finbuckle.MultiTenant\\" Version=\\"6.11.1\\"/>
  </ItemGroup>
</Project>
"
`;
